VMware: Create vmware_host_kernel_manager.py

alongchamps opened the pull request:
Implements vmware_host_kernel_manager and provides code for #52079.

##### SUMMARY
This implements a module to configure ESXi kernel module options.

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request

##### COMPONENT NAME
vmware_host_kernel_manager

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
This is a new module I created as an exercise to test how easy it was to create a module for VMware ESXi hosts. This also provides functionality that was not available currently in the VMware scripts.
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below

```

alongchamps committed:
Create vmware_host_kernel_manager.py

Initial push. Implements vmware_host_kernel_manager and provides code for 52079 on the ansible project.

alongchamps committed:
Update vmware_host_kernel_manager.py

Fixed typo on kernel_module_option in documentation block.

alongchamps committed:
Update vmware_host_kernel_manager.py

Fixing linting errors including examples, spacing, and returns.

alongchamps committed:
Update vmware_host_kernel_manager.py

Added required imports on top.

alongchamps committed:
Update vmware_host_kernel_manager.py

Fixing spaces and adjusting comments.

alongchamps committed:
Update vmware_host_kernel_manager.py

This push goes out to the newline x2.

alongchamps commented:
I'm not sure if someone in the project can squash my commits down but in the libertymutual repo it came from, I don't have those permissions it looks like and my contact is out of the office.

alongchamps commented:
cc @dagwieers - we talked about this module last week on IRC so that's why I tagged you on this. It seems like all the tests are passing on shippable and looks good.

Akasurde commented:
@alongchamps Thanks for the contribution. I will squash and merge the changes so don't worry about squash.

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_host_kernel_manager.py:
```diff
@@ -0,0 +1,221 @@
+    Users can specify an ESXi hostname or cluster name. All connected ESXi hosts in scope will be
+    configured when specified. If a host is not connected at time of configuration, it will be marked
+    as such in the output and be skipped. Kernel module options may require a reboot to take effect
+    which is not covered by this script.
```
You can add a suggestion - 

Kernel module options may require a reboot to take effect. You can use M(reboot) or M(vmware_host_powerstate) module to reboot all ESXi host systems.

dagwieers replied:
Please make this different items in the description.

alongchamps replied:
Comments have been adjusted for this.

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_host_kernel_manager.py:
```diff
@@ -0,0 +1,221 @@
+options:
+  esxi_hostname:
+    description:
+    - Name of the ESXi host to work on
```
```suggestion
    - Name of the ESXi host to work on.
```

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_host_kernel_manager.py:
```diff
@@ -0,0 +1,221 @@
+    type: str
+  kernel_module_name:
+    description:
+    - Name of the kernel module to be configured
```
```suggestion
    - Name of the kernel module to be configured.
```

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_host_kernel_manager.py:
```diff
@@ -0,0 +1,221 @@
+    - Specified configurations will be applied to the given module
+    - These values are specified in key=value pairs and separated by a space when there are multiple options
+    required: true
+    type: str
```
how about creating  a list of key and value pairs 

```yaml
- name: Configure kernel module parameters
  vmware_host_kernel_manager:
    hostname: '{{ vcenter_hostname }}'
    username: '{{ vcenter_username }}'
    password: '{{ vcenter_password }}'
    esxi_hostname: '{{ esxi_hostname }}'
    kernel_modules:
      - "tcpip4": "ipv6=0"
      - "vmw_psp_rr": "maxPathsPerDevice=2"
```

alongchamps replied:
Talked with akasurde on IRC, for now I'm going to leave it as kernel_module_name and kernel_module_option because it can be looped over with loop controls built in to ansible already.

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_host_kernel_manager.py:
```diff
@@ -0,0 +1,221 @@
+'''
+
+try:
+    from pyVmomi import vim, vmodl
```
vmodl is unused.

alongchamps replied:
Removed.

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_host_kernel_manager.py:
```diff
@@ -0,0 +1,221 @@
+
+    # find kernel module options for a given kmod_name. If the name is not right, this will throw an exception
+    def get_kernel_module_option(self, host, kmod_name):
+        host_kernel_manager = host.configManager.kernelModuleSystem
```
Sometimes `ModuleSystem` is None so 
```python
if host_kernel_manager is None:
  self.module.fail_json(msg="Failed to find Kernel module system for %s" % host.name)
```

alongchamps replied:
Talked to akasurde on IRC, this is already checked in check_host_configuration_state which is called first before the get_kernel_module_option and apply_kernel_module_option functions.

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_host_kernel_manager.py:
```diff
@@ -0,0 +1,221 @@
+        host_kernel_manager = host.configManager.kernelModuleSystem
+
+        try:
+            configured_module_options = host_kernel_manager.QueryConfiguredModuleOptionString(self.kernel_module_name)
```
```suggestion
            return host_kernel_manager.QueryConfiguredModuleOptionString(self.kernel_module_name)
```

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_host_kernel_manager.py:
```diff
@@ -0,0 +1,221 @@
+            try:
+                if not self.module.check_mode:
+                    host_kernel_manager.UpdateModuleOptionString(kmod_name, kmod_option)
+                    self.results[host.name]['changed'] = True
```
```suggestion
                 self.results[host.name]['changed'] = True
```

Akasurde replied:
Want to report changed as True even if module is ran under `check_mode`

alongchamps replied:
Ok, that is good to know.

I'm going to actually remove line 141 above because I have the changed status being assigned inside check_host_configuration_state

alongchamps replied:
I think this one is going to be resolved in the next commit akasurde. We can touch base tomorrow on this one in particular.

Akasurde reviewed (changes requested):
Could you please add an integration test for this change ? I am not aware that VCSIM does not support `kernelModuleSystem` but you can write testcase and mark it as unsupported like [vmware_guest_snapshot_facts](https://github.com/ansible/ansible/blob/devel/test/integration/targets/vmware_guest_snapshot_facts/aliases).

dagwieers commented on lib/ansible/modules/cloud/vmware/vmware_host_kernel_manager.py:
```diff
@@ -0,0 +1,221 @@
+  esxi_hostname:
+    description:
+    - Name of the ESXi host to work on
+    - This parameter is required if C(cluster_name) is not specified
```
```suggestion
    - This parameter is required if C(cluster_name) is not specified.
```

dagwieers replied:
Please make sure all descriptions have a trailing dot.

dagwieers commented on lib/ansible/modules/cloud/vmware/vmware_host_kernel_manager.py:
```diff
@@ -0,0 +1,221 @@
+extends_documentation_fragment: vmware.documentation
+'''
+
+EXAMPLES = '''
```
```suggestion
EXAMPLES = r'''
```

dagwieers commented on lib/ansible/modules/cloud/vmware/vmware_host_kernel_manager.py:
```diff
@@ -0,0 +1,221 @@
+        ],
+        mutually_exclusive=[
+            ['cluster_name', 'esxi_hostname'],
+        ]
```
```suggestion
        ],
```

alongchamps replied:
Added. Still learning python syntax.

dagwieers commented on lib/ansible/modules/cloud/vmware/vmware_host_kernel_manager.py:
```diff
@@ -0,0 +1,221 @@
+module: vmware_host_kernel_manager
+short_description: Configure kernel module options on an ESXi host or cluster.
+description:
+>-
```
Don't do this, make these items in your description-list.

alongchamps replied:
Comment has been adjusted.

dagwieers commented on lib/ansible/modules/cloud/vmware/vmware_host_kernel_manager.py:
```diff
@@ -0,0 +1,221 @@
+DOCUMENTATION = r'''
+---
+module: vmware_host_kernel_manager
+short_description: Configure kernel module options on an ESXi host or cluster.
```
Drop the trailing dot for short_description.
```suggestion
short_description: Manage kernel module options on ESXi hosts
```

dagwieers reviewed (changes requested):
I don't know if this is a thing, but having the module make changes for a cluster of hosts is no doubt more efficient, but is contrary to how things work in Ansible. Normally your targets would be ESXi hosts, not your cluster. So I don't know what the normal pattern is, and we should have documentation detailing how people are managing their VMware infrastructure. One reason is that returning information per host breaks how you would expect this to work on a per-host basis.

Also, I think we need to have integration tests for *every* module. Maybe the integration tests may need to be disabled by default because we cannot test them in Shippable CI, but at least we can guarantee that the module will behave as-is when future changes are made. So it's a safety net if you depend on the current behavior.

alongchamps committed:
Updating with fixes that have been suggested

Many updates for formatting, descriptions, and some minor adjustments for particular values such as 'changed'. Also added one integration tests (albeit unsupported in govcsim). I have more to add, but I will see how this goes for now.

alongchamps committed:
Update vmware_host_kernel_manager.py

Found a stray semicolon...

alongchamps committed:
Update main.yml

Fixing main.yml syntax.

alongchamps commented:
Previous test failed with an error in aws/3.6/2:
00:00 Error: No such container: 0b8108f5829e

alongchamps committed:
Update main.yml

Added another test to trigger shippable again.

alongchamps commented:
Hi @Akasurde and @dagwieers, I think I have implemented all the requested changes in my latest commit.

@dagwieers for your comment on the cluster/single ESXi host stuff, I think that might be a good topic for the next #ansible-vmware meeting next week so we're all on the same page for how the VMware modules go. The cluster and single host options are mutually exclusive, so whatever target is set is in scope.

dagwieers commented:
@alongchamps Sure, I am not really involved in VMware development, so maybe there's no issue and this is just common practice. It only looked weird for me, but I do see the value.

Akasurde commented:
@alongchamps @dagwieers Most `vmware_host_*` implement same logic as this module does. People have different approaches to handle host level configurations. Some uses host level configurations and some uses cluster to apply setting to all ESXi host systems.

Akasurde reviewed (approved):
LGTM

alongchamps committed:
Merge pull request #1 from ansible/devel

Resync with ansible/ansible/devel branch latest changes

alongchamps committed:
Revert "Resync with ansible/ansible/devel branch latest changes"

alongchamps committed:
Merge pull request #2 from libertymutual/revert-1-devel

Revert "Resync with ansible/ansible/devel branch latest changes"

Akasurde commented:
superseded by https://github.com/ansible/ansible/pull/52503
